### ISSUE TITLE
Memory leak in model.fit

### ISSUE THREAD

**Issue by krenerd** [tensorflow/tensorflow#50379]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

The following issue still remains to be prevailing regardless of TF versions for many people, and I ought to re-open this issue on memory leakage. 

https://github.com/tensorflow/tensorflow/issues/37505

Currently, the only solution is a crack to wrap the sequence into multiprocessing. I hope this bug could be fixed as soon as possible.

**Comment by tilakrayal** [comment]
@krenerd ,

In order to expedite the trouble-shooting process, could you please provide the tensorflow version,complete code and dataset to reproduce the issue reported here.

**Comment by krenerd** [comment]
> System information

Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
minimal working example

OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
windows-server 2016

TensorFlow installed from (source or binary):
conda

TensorFlow version (use command below):
tf 2.1.0

Python version:
3.7

CUDA/cuDNN version:
CUDA 10.1

GPU model and memory:
K80 - 24Gb

Describe the current behavior
memory use increases with consecutive training runs, probably related to #35524, #33030, #35124, #35835
side note: I do not understand the warning but this seems to be handled in #37500
Describe the expected behavior
memory should stay constant

Standalone code to reproduce the issue
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
```
from tensorflow.keras.datasets import cifar10
import tensorflow.keras.callbacks as callbacks
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.callbacks import LearningRateScheduler
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.layers import Input, Conv2D, GlobalAveragePooling2D, Activation, Dense
from tensorflow.keras.models import Model
import tensorflow.keras.utils as kutils
import numpy as np
import psutil
import gc 


batch_size = 128
epochs = 5
num_classes = 10


def buildmodel():
    img_input = Input(shape=(32, 32, 3))

    x = Conv2D(16, (3, 3), padding='same')(img_input)
    x = Activation("relu")(x)
    x = Conv2D(16, (3, 3), padding='same')(x)
    x = Activation("relu")(x)
    x = GlobalAveragePooling2D()(x)
    prediction = Dense(num_classes,activation='softmax', name = 'classifier') (x)  
    model = Model(inputs=img_input, outputs=prediction)
    return model

(trainX, trainY), (testX, testY) = cifar10.load_data()
mean = np.mean(trainX, axis=0)
std = np.std(trainX)

trainX = trainX.astype('float32')
trainX = (trainX - mean) / std
testX = testX.astype('float32')
testX = (testX - mean) / std

trainY = kutils.to_categorical(trainY)
testY = kutils.to_categorical(testY)

generator = ImageDataGenerator()
generator.fit(trainX)
val_generator = ImageDataGenerator()

for i in range(10):
    tf.keras.backend.clear_session()
    model = buildmodel()
    sgd = SGD(lr=0.1, momentum=0.9, nesterov=True)
    model.compile(loss="categorical_crossentropy", optimizer=sgd, metrics=["acc"])
    model.fit(generator.flow(trainX, trainY, batch_size=batch_size), epochs=epochs, validation_data=val_generator.flow(testX, testY, batch_size = batch_size),verbose=0, workers = 20)
    print('memory usesd: ' + str(psutil.virtual_memory().used // 1e6))
    gc.collect()
```
output:
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
memory usesd: 38012.0
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
memory usesd: 38563.0
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
memory usesd: 39288.0
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
memory usesd: 40005.0
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
memory usesd: 40730.0
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
memory usesd: 41490.0
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
memory usesd: 42216.0
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
memory usesd: 42937.0
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
memory usesd: 43659.0
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
WARNING:tensorflow:sample_weight modes were coerced from
...
to
['...']
memory usesd: 44403.0

**Comment by tilakrayal** [comment]
@krenerd ,

Could you please update TensorFlow to the latest stable version v2.5 and check if you are facing the same issue. Thanks!

**Comment by krenerd** [comment]
The situation is that I previously experienced this issue about 5 months ago and while I was currently working with TF 2.4 on another project, I  had this issue again. So I didn't test anything on v2.5 but it seems that other people including @jayagami was experiencing this issue even for TF 2.5 according to the threads in https://github.com/tensorflow/tensorflow/issues/37505. 

So this doesn't seem like a version issue and the code above will still reproduce the leak in TF 2.5.

**Comment by iksooman** [comment]
I also met the memory issue when I used @tf.function.
`
def train_step():
     @tf.function
     def train_step1(data, model, lr, loss_object, optimizer, train_loss, train_accuracy):
         with tf.GradientTape() as tape:
             prediction = model(data[0])
             loss = loss_object(data[1], prediction)

         gradients = tape.gradient(loss, model.trainable_variables)
         optimizer.learning_rate = lr
         optimizer.apply_gradients(zip(gradients, model.trainable_variables))

         train_loss(loss)
         train_accuracy(data[1], prediction)

     return train_step1
`
After deleting @tf.function, it doesn't make the issue.

**Comment by iksooman** [comment]
I also met the memory issue when I used @tf.function.

`def train_step():
     @tf.function
     def train_step1(data, model, lr, loss_object, optimizer, train_loss, train_accuracy):
     ¦   with tf.GradientTape() as tape:
     ¦   ¦   prediction = model(data[0])
     ¦   ¦   loss = loss_object(data[1], prediction)

     ¦   gradients = tape.gradient(loss, model.trainable_variables)
     ¦   optimizer.learning_rate = lr
     ¦   optimizer.apply_gradients(zip(gradients, model.trainable_variables))

     ¦   train_loss(loss)
     ¦   train_accuracy(data[1], prediction)

     return train_step1`

After deleting @tf.function, it doesn't make the issue.

**Comment by iksooman** [comment]
I also met the memory issue when I used @tf.function

```
def train_step():
     @tf.function
     def train_step1(data, model, lr, loss_object, optimizer, train_loss, train_accuracy)
         with tf.GradientTape() as tape:
             prediction = model(data[0])
             loss = loss_object(data[1], prediction)

         gradients = tape.gradient(loss, model.trainable_variables)
         optimizer.learning_rate = lr
         optimizer.apply_gradients(zip(gradients, model.trainable_variables))

         train_loss(loss)
         train_accuracy(data[1], prediction)

     return train_step1
```

After deleting @tf.function, it doesn't make the issue any more.
(tensorflow==2.4.0)

**Comment by tilakrayal** [comment]
@krenerd ,

I ran your code with tf-nightly and I don't see any improvement in the memory as much as you have noticed. When I ran your code with tf-nightly, i see the following ouput.

**Comment by krenerd** [comment]
Cool, I understand that the bug was fixed in `tf-nightly` since I also tried it in my local environment w/o memory leaks.

**Comment by dcbarton** [comment]
This memory leak is most definitely still not fixed. Keras remains unusable for any longer training time unless a custom raining loop is used.

**Comment by dcbarton** [comment]
After months dealing with this unbelievable memory leak, I think I stumbled on a simple workaround -- return sample weights from the generator. I introduced them today for other reasons and the memory leak completely disappeared. I've never been so happy in my life, I'm going to buy a lotto ticket.

TF folks really need to fix this, but I hope others confirm this simple workaround.

**Comment by tilakrayal** [comment]
Hello @dcbarton ,
As this issue is closed, I request to please submit a new issue from [this link](https://github.com/tensorflow/tensorflow/issues/new/choose) and fill in the template, so that we can track the issue there. Thanks!

**Comment by noamelata** [comment]
> I explored further and a separate, trivial change was responsible for curing the memory leak. In on_epoch_end I was updaing a generator class variable with a new numpy array. The old numpy arrays should be collected by normal gc cycles, but never are. They pile up, and even forcing gc will not collect them. I changed the logic to alter the array in place instead of creating a new one, and the memory leak is gone.
> 
> I hope this helps someone at TF solve this bug. I also hope this helps others with terrible memory leaks in Keras.

Hi @dcbarton ,
Can you please post the code responsible for the fix? 
It would be a great help

**Comment by dcbarton** [comment]
THIS CODE ACCUMULATES MEMORY LEAK, EVEN WITH FORCED GC
class DataGenerator(Sequence):
    def on_epoch_end(self):
        self.indexes = np.arange(self.startIdx, self.endIdx, 1, dtype=np.int32)
        np.random.shuffle(self.indexes)

THIS CODE WORKS AROUND BY REUSING THE BUFFER, BETTER CODE ANYWAY
class DataGenerator(Sequence):
    def __init__(self):
        self.indexes = np.arange(self.startIdx, self.endIdx, 1, dtype=np.int32)
    def on_epoch_end(self):        
        np.random.shuffle(self.indexes)

**Comment by noamelata** [comment]
Thank you, will update if this helped
